pkg/importer/data-processor.go

```diff
@@ -183,66 +195,93 @@ func (dp *DataProcessor) ProcessDataResume() error {
+	})
+	dp.RegisterPhaseExecutor(ProcessingPhaseConvert, func() (ProcessingPhase, error) {
+		pp, err := dp.convert(dp.source.GetURL())
+		if err != nil {
+			err = errors.Wrap(err, "Unable to convert source data to target format")
+		}
+		return pp, err
+	})
+	dp.RegisterPhaseExecutor(ProcessingPhaseResize, func() (ProcessingPhase, error) {
+		pp, err := dp.resize()
+		if err != nil {
+			err = errors.Wrap(err, "Unable to resize disk image to requested size")
+		}
+		return pp, err
+	})
+	dp.RegisterPhaseExecutor(ProcessingPhaseMergeDelta, func() (ProcessingPhase, error) {
+		pp, err := dp.merge()
+		if err != nil {
+			err = errors.Wrap(err, "Unable to apply delta to base image")
+		}
+		return pp, err
+	})
+}
+
 // ProcessDataWithPause is the main processing loop.
 func (dp *DataProcessor) ProcessDataWithPause() error {
-	var err error
+	visited := make(map[ProcessingPhase]bool, len(dp.phaseExecutors))
 	for dp.currentPhase != ProcessingPhaseComplete && dp.currentPhase != ProcessingPhasePause {
-		switch dp.currentPhase {
-		case ProcessingPhaseInfo:
-			dp.currentPhase, err = dp.source.Info()
-			if err != nil {
-				err = errors.Wrap(err, "Unable to obtain information about data source")
-			}
-		case ProcessingPhaseTransferScratch:
-			dp.currentPhase, err = dp.source.Transfer(dp.scratchDataDir)
-			if err == ErrInvalidPath {
-				// Passed in invalid scratch space path, return scratch space needed error.
-				err = ErrRequiresScratchSpace
-			} else if err != nil {
-				err = errors.Wrap(err, "Unable to transfer source data to scratch space")
-			}
-		case ProcessingPhaseTransferDataDir:
-			dp.currentPhase, err = dp.source.Transfer(dp.dataDir)
-			if err != nil {
-				err = errors.Wrap(err, "Unable to transfer source data to target directory")
-			}
-		case ProcessingPhaseTransferDataFile:
-			dp.currentPhase, err = dp.source.TransferFile(dp.dataFile)
-			if err != nil {
-				err = errors.Wrap(err, "Unable to transfer source data to target file")
-			}
-		case ProcessingPhaseValidatePause:
-			validateErr := dp.validate(dp.source.GetURL())
-			if validateErr != nil {
-				dp.currentPhase = ProcessingPhaseError
-				err = validateErr
-			}
-			dp.currentPhase = ProcessingPhasePause
-		case ProcessingPhaseConvert:
-			dp.currentPhase, err = dp.convert(dp.source.GetURL())
-			if err != nil {
-				err = errors.Wrap(err, "Unable to convert source data to target format")
-			}
-		case ProcessingPhaseResize:
-			dp.currentPhase, err = dp.resize()
-			if err != nil {
-				err = errors.Wrap(err, "Unable to resize disk image to requested size")
-			}
-		case ProcessingPhaseMergeDelta:
-			dp.currentPhase, err = dp.merge()
-			if err != nil {
-				err = errors.Wrap(err, "Unable to apply delta to base image")
-			}
-		default:
+		if visited[dp.currentPhase] {
+			err := errors.Errorf("loop detected on phase %s", dp.currentPhase)
+			klog.Errorf("%+v", err)
+			return err
+		}
+		executor, ok := dp.phaseExecutors[dp.currentPhase]
+		if !ok {
```

Comment:
This branch is not covered by test. It would be good to write a test for a missing/unknown phase, so it is protected from future refactorings.

Reply:
Test case was added in the last commit.

Thank you for the review!